processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java

```diff
@@ -143,10 +142,11 @@ private InDimFilter(
 
     // The values set can be huge. Try to avoid copying the set if possible.
     // Note that we may still need to copy values to a list for caching. See getCacheKey().
-    if ((NullHandling.sqlCompatible() || values.stream().noneMatch(NullHandling::needsEmptyToNull))) {
+    if (NullHandling.sqlCompatible() || !values.remove("")) {
       this.values = values;
     } else {
-      this.values = values.stream().map(NullHandling::emptyToNullIfNeeded).collect(Collectors.toSet());
+      values.add(null);
```

Comment:
Nice finding, but this can lead to incorrect result in default mode since `null` will be always added no matter whether an empty string is in `values` or not.

Reply:
I think the if condition on line 145 works such that it only enters this else block if there was an empty string already in the set. I should add unit tests here, because I had to think a lot about how the ordering of the if statements affected the code flow 😅 and how nulls and empty strings work in the different modes...

What I want to happen here is:
* If it's sqlCompatible mode, just use the values as is
* If it's default mode (ie not sql compatible mode) attempt to remove empty string
* If an empty string was removed add null
* If no empty  string was removed, use values as is

Reply:
Oh sorry, you're right. I misread the code. It would be nice to add some comment about what we want to do here since it's not much intuitive.

Reply:
In most likelihood, the `values` is a `HashSet` which is internally backed by a `HashMap` and hence `remove, `contains` will be much faster than linear scan. 

```suggestion
      if ((NullHandling.sqlCompatible()) {
      this.values = values;
    } else if (values.remove("")) {
        values.add(null);
    }
```

Reply:
I've re-written the if statement to hopefully be easier to follow.